node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+
+//! The bitfield distribution
+//!
+//! In case this node is a validator, gossips its own signed availability bitfield
+//! for a particular relay parent.
+//! Independently of that, gossips on received messages from peers to other interested peers.
+
+use codec::{Decode, Encode};
+use futures::{channel::oneshot, FutureExt};
+
+use keystore::KeyStorePtr;
+use sc_keystore as keystore;
+
+use node_primitives::{ProtocolId, View};
+
+use log::{trace, warn};
+use polkadot_erasure_coding::{
+	branch_hash, branches, obtain_chunks_v1 as obtain_chunks};
+use polkadot_primitives::v1::{
+	AvailableData, BlakeTwo256, CommittedCandidateReceipt, CoreState, ErasureChunk,
+	GlobalValidationData, Hash as H256, HashT, Id as ParaId, LocalValidationData,
+	OccupiedCoreAssumption, OmittedValidationData, PoV,
+	ValidatorId, ValidatorIndex, ValidatorPair,
+};
+use polkadot_subsystem::messages::*;
+use polkadot_subsystem::{
+	errors::{ChainApiError, RuntimeApiError}, ActiveLeavesUpdate, FromOverseer, OverseerSignal, SpawnedSubsystem,
+	Subsystem, SubsystemContext, SubsystemError,
+};
+use sc_network::ReputationChange as Rep;
+use std::collections::{HashMap, HashSet};
+use std::io;
+use std::iter;
+
+const TARGET: &'static str = "avad";
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	#[from]
+	Erasure(polkadot_erasure_coding::Error),
+	#[from]
+	Io(io::Error),
+	#[from]
+	Oneshot(oneshot::Canceled),
+	#[from]
+	Subsystem(SubsystemError),
+	#[from]
+	RuntimeApi(RuntimeApiError),
+	#[from]
+	ChainApi(ChainApiError),
+
+	Logic,
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+const COST_MERKLE_PROOF_INVALID: Rep = Rep::new(-100, "Merkle proof was invalid");
+const COST_NOT_A_LIVE_CANDIDATE: Rep = Rep::new(-51, "Candidate is not live");
+const COST_MESSAGE_NOT_DECODABLE: Rep = Rep::new(-100, "Message is not decodable");
+const COST_PEER_DUPLICATE_MESSAGE: Rep = Rep::new(-500, "Peer sent identical messages");
+const BENEFIT_VALID_MESSAGE_FIRST: Rep = Rep::new(15, "Valid message with new information");
+const BENEFIT_VALID_MESSAGE: Rep = Rep::new(10, "Valid message");
+
+/// Checked signed availability bitfield that is distributed
+/// to other peers.
+#[derive(Encode, Decode, Debug, Clone, PartialEq, Eq, Hash)]
+pub struct AvailabilityGossipMessage {
+	/// Ankor hash of the candidate this is associated to.
+	pub candidate_hash: H256,
+	/// The actual signed availability bitfield.
+	pub erasure_chunk: ErasureChunk,
+}
+
+/// Data used to track information of peers and relay parents the
+/// overseer ordered us to work on.
+#[derive(Default, Clone)]
+struct ProtocolState {
+	/// Track all active peers and their views
+	/// to determine what is relevant to them.
+	peer_views: HashMap<PeerId, View>,
+
+	/// Our own view.
+	view: View,
+
+	/// Caches a mapping of relay parents to live candidates
+	/// and allows fast + intersection free obtaining of active heads set by unionizing.
+	/// Includes the `K` ancestors too.
```

Comment:
As in, this includes an entry for each of the `K` ancestors, or this contains an entry for each relay-parent that includes the candidates for each of the `K` ancestors?

The first option is preferable and the comment below leads me to believe it is what is implemented

Reply:
That is correct, each relay parent as well as each ancestor hash is a key with an associated receipt each.